PublishedModules/Psychokiller1888/Calculator/Calculator.py

```diff
@@ -0,0 +1,83 @@
+import math
+
+from core.base.model.Intent import Intent
+from core.base.model.Module import Module
+from core.dialog.model.DialogSession import DialogSession
+
+
+class Calculator(Module):
+	"""
+	Author: Psychokiller1888
+	Description: Do some calculation with alice
+	"""
+
+	_INTENT_MATHS = Intent('Maths')
+
+
+	def __init__(self):
+		self._SUPPORTED_INTENTS = [
+			self._INTENT_MATHS
+		]
+		self._lastNumber = None
+		super().__init__(self._SUPPORTED_INTENTS)
+
+
+	def onMessage(self, intent: str, session: DialogSession) -> bool:
+		if not self.filterIntent(intent, session):
+			return False
+
+		slots = session.slotsAsObjects
+		sessionId = session.sessionId
+
+		if intent == self._INTENT_MATHS:
+			if ('Left' not in slots and 'Right' not in slots) or 'Function' not in slots:
+				self.continueDialog(sessionId=sessionId, text=self.TalkManager.randomTalk('notUnderstood'))
+				return True
+
+			func = slots['Function'][0].value['value']
+
+			if 'Left' in slots and 'Right' in slots:
+				result = self.calculate(float(slots['Left'][0].value['value']), float(slots['Right'][0].value['value']), func)
+
+			elif 'Right' in slots and 'Left' not in slots:
+				if not self._lastNumber:
+					self.endDialog(sessionId=sessionId, text=self.TalkManager.randomTalk('noPreviousOperation'))
+					return True
+
+				result = self.calculate(self._lastNumber, float(slots['Right'][0].value['value']), func)
+
+			else:
+				self.continueDialog(sessionId=sessionId, text=self.TalkManager.randomTalk('notUnderstood'))
+				return True
+
+			answer = str(int(result)) if result % 1 == 0 else str(result)
+
+			self.endDialog(sessionId=sessionId, text=answer)
+
+		return True
+
+
+	def calculate(self, left: float, right: float, func: str) -> float:
+		if func == '+':
+			result = left + right
+		elif func == '-':
+			result = left - right
+		elif func == '/':
+			result = left / right
+		elif func == '*':
+			result = left * right
+		elif func == 'square root':
+			result = math.sqrt(left)
+		elif func == 'modulo':
+			result = left % right
+		elif func == 'sine':
+			result = round(math.sin(left), 3)
+		elif func == 'cosine':
+			result = round(math.cos(left), 3)
+		elif func == 'tangent':
+			result = round(math.tan(left), 3)
+		else:
+			result = 'not supported'
```

Comment:
this return value is not of type float and self._lastNumber should not be a string because then this function gets called with a string the next time which causes string + int -> TypeError

Maybe set result to None when there is no result and then use
```
answer = str(int(result)) if result == 0 else 'not supported'
```
in line 53?

Reply:
Problem casting to int you loose the decimals on not full divisions

Reply:
oh did not read it correctly xD that has to stay in there, but still can output the sentence when it is None

Reply:
```
if result is None:
    answer = 'not supported'
elif result % 1 == 0:
  answer = str(int(result))
else:
  answer = str(result)
```

Reply:
yep, better